runtime/parachains/src/scheduler.rs

```diff
@@ -483,77 +507,102 @@ impl<T: Config> Pallet<T> {
 			let cq = ClaimQueue::<T>::get();
 
 			for core_idx in 0..n_session_cores {
-				let core_idx = CoreIndex(core_idx);
-				let group_idx = Self::group_assigned_to_core(core_idx, now).expect(
-					"core is not out of bounds and we are guaranteed \
-										  to be after the most recent session start; qed",
-				);
+				let core_idx = CoreIndex::from(core_idx);
 
 				// add previously timedout paras back into the queue
-				if let Some(assignment) = timedout_paras.remove(&core_idx) {
-					let ca = assignment.to_core_assignment(core_idx, group_idx);
-					Self::add_to_claimqueue(ca);
+				if let Some(mut entry) = timedout_paras.remove(&core_idx) {
+					if entry.retries < T::AssignmentProvider::get_max_retries(core_idx) {
+						entry.retries += 1;
+						Self::add_to_claimqueue(core_idx, entry);
+					} else {
+						// Consider max retried parathreads as concluded for the assignment provider
+						debug_assert!(concluded_paras.insert(core_idx, entry.para_id()).is_none());
```

Comment:
This is no good. insert has a side effect of inserting that is intended even in release builds.

Reply:
Oh wow! I didn't expect `debug_assert` to be completely removed in release builds! I thought it would turn the panicky assert into a log instead :facepalm: